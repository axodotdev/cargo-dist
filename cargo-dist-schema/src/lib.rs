#![deny(missing_docs)]

//! # cargo-dist-schema
//!
//! This crate exists to serialize and deserialize the dist-manifest.json produced
//! by cargo-dist. Ideally it should be reasonably forward and backward compatible
//! with different versions of this format.
//!
//! The root type of the schema is [`DistManifest`][].

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// A local system path on the machine cargo-dist was run.
///
/// This is a String because when deserializing this may be a path format from a different OS!
pub type LocalPath = String;
/// A relative path inside an artifact
///
/// This is a String because when deserializing this may be a path format from a different OS!
///
/// (Should we normalize this one?)
pub type RelPath = String;

/// A report of the releases and artifacts that cargo-dist generated
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DistManifest {
    /// The version of cargo-dist that generated this
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dist_version: Option<String>,
    /// App releases we're distributing
    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub releases: Vec<Release>,
}

/// A Release of an Application
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Release {
    /// The name of the app
    pub app_name: String,
    /// The version of the app
    // FIXME: should be a Version but JsonSchema doesn't support (yet?)
    pub app_version: String,
    /// The artifacts for this release (zips, debuginfo, metadata...)
    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub artifacts: Vec<Artifact>,
    /// The title of the changelog for this release
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub changelog_title: Option<String>,
    /// The body of the changelog for this release
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub changelog_body: Option<String>,
}

/// A distributable artifact that's part of a Release
///
/// i.e. a zip or installer
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Artifact {
    /// The unique name of the artifact (e.g. `myapp-v1.0.0-x86_64-pc-windows-msvc.zip`)
    ///
    /// If this is missing then that indicates the artifact is purely informative and has
    /// no physical files associated with it. This may be used (in the future) to e.g.
    /// indicate you can install the application with `cargo install` or `npm install`.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub name: Option<String>,
    /// The kind of artifact this is (e.g. "exectuable-zip")
    #[serde(flatten)]
    pub kind: ArtifactKind,
    /// The target triple of the bundle
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(default)]
    pub target_triples: Vec<String>,
    /// The location of the artifact on the local system
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub path: Option<LocalPath>,
    /// Assets included in the bundle (like executables and READMEs)
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(default)]
    pub assets: Vec<Asset>,
    /// A string describing how to install this
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub install_hint: Option<String>,
    /// A brief description of what this artifact is
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub description: Option<String>,
}

/// An asset contained in an artifact (executable, license, etc.)
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Asset {
    /// The high-level name of the asset
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// The path of the asset relative to the root of the artifact
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<RelPath>,
    /// The kind of asset this is
    #[serde(flatten)]
    pub kind: AssetKind,
}

/// An artifact included in a Distributable
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind")]
#[non_exhaustive]
pub enum AssetKind {
    /// An executable artifact
    #[serde(rename = "executable")]
    Executable(ExecutableAsset),
    /// A README file
    #[serde(rename = "readme")]
    Readme,
    /// A LICENSE file
    #[serde(rename = "license")]
    License,
    /// A CHANGELOG or RELEASES file
    #[serde(rename = "changelog")]
    Changelog,
    /// Unknown to this version of cargo-dist-schema
    ///
    /// This is a fallback for forward/backward-compat
    #[serde(other)]
    #[serde(rename = "unknown")]
    Unknown,
}

/// A kind of Artifact
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind")]
#[non_exhaustive]
pub enum ArtifactKind {
    /// A zip or a tarball
    #[serde(rename = "executable-zip")]
    ExecutableZip,
    /// Standalone Symbols/Debuginfo for a build
    #[serde(rename = "symbols")]
    Symbols,
    /// Machine-readable metadata
    #[serde(rename = "dist-metadata")]
    DistMetadata,
    /// Installer
    #[serde(rename = "installer")]
    Installer,
    /// Unknown to this version of cargo-dist-schema
    ///
    /// This is a fallback for forward/backward-compat
    #[serde(other)]
    #[serde(rename = "unknown")]
    Unknown,
}

/// An executable artifact (exe/binary)
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ExecutableAsset {
    /// The name of the Artifact containing symbols for this executable
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub symbols_artifact: Option<String>,
}

impl DistManifest {
    /// Create a new DistManifest
    pub fn new(dist_version: String, releases: Vec<Release>) -> Self {
        Self {
            dist_version: Some(dist_version),
            releases,
        }
    }

    /// Get the JSON Schema for a DistManifest
    pub fn json_schema() -> schemars::schema::RootSchema {
        schemars::schema_for!(DistManifest)
    }
}

#[test]
fn emit() {
    use std::fs::File;
    use std::io::BufWriter;
    use std::io::Write;
    use std::path::PathBuf;

    let schema = DistManifest::json_schema();
    let json_schema = serde_json::to_string_pretty(&schema).unwrap();
    insta::assert_snapshot!(json_schema);

    // FIXME: (?) we should use something like xtask to update the schema, but this works ok.
    let root = std::env!("CARGO_MANIFEST_DIR");
    let schema = PathBuf::from(root).join("cargo-dist-json-schema.json");
    let file = File::options()
        .create(true)
        .write(true)
        .truncate(true)
        .open(schema)
        .unwrap();
    let mut file = BufWriter::new(file);
    writeln!(&mut file, "{json_schema}").unwrap();
}
