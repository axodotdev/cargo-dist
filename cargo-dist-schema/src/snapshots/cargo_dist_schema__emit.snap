---
source: cargo-dist-schema/src/lib.rs
expression: json_schema
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "DistManifest",
  "description": "A report of the releases and artifacts that dist generated",
  "type": "object",
  "properties": {
    "announcement_changelog": {
      "description": "A changelog for the announcement",
      "type": [
        "string",
        "null"
      ]
    },
    "announcement_github_body": {
      "description": "A Github Releases body for the announcement",
      "type": [
        "string",
        "null"
      ]
    },
    "announcement_is_prerelease": {
      "description": "Whether this announcement appears to be a prerelease",
      "default": false,
      "type": "boolean"
    },
    "announcement_tag": {
      "description": "The (git) tag associated with this announcement",
      "type": [
        "string",
        "null"
      ]
    },
    "announcement_tag_is_implicit": {
      "description": "True if --tag wasn't explicitly passed to dist. This usually indicates some kind of dry-run state like pr-run-mode=upload. Some third-party tools may use this as a proxy for \"is dry run\"",
      "default": false,
      "type": "boolean"
    },
    "announcement_title": {
      "description": "A title for the announcement",
      "type": [
        "string",
        "null"
      ]
    },
    "artifacts": {
      "description": "The artifacts included in this Announcement, referenced by releases.",
      "type": "object",
      "additionalProperties": {
        "$ref": "#/definitions/Artifact"
      }
    },
    "assets": {
      "description": "The assets contained within artifacts (binaries)",
      "type": "object",
      "additionalProperties": {
        "$ref": "#/definitions/AssetInfo"
      }
    },
    "ci": {
      "description": "ci backend info",
      "anyOf": [
        {
          "$ref": "#/definitions/CiInfo"
        },
        {
          "type": "null"
        }
      ]
    },
    "dist_version": {
      "description": "The version of dist that generated this",
      "type": [
        "string",
        "null"
      ]
    },
    "force_latest": {
      "description": "Where possible, announce/publish a release as \"latest\" regardless of semver version",
      "default": false,
      "type": "boolean"
    },
    "github_attestations": {
      "description": "Whether Artifact Attestations should be found in the GitHub Release\n\n<https://github.blog/2024-05-02-introducing-artifact-attestations-now-in-public-beta/>",
      "type": "boolean"
    },
    "linkage": {
      "description": "Data about dynamic linkage in the built libraries",
      "default": [],
      "type": "array",
      "items": {
        "$ref": "#/definitions/Linkage"
      }
    },
    "publish_prereleases": {
      "description": "Whether to publish prereleases to package managers",
      "default": false,
      "type": "boolean"
    },
    "releases": {
      "description": "App releases we're distributing",
      "type": "array",
      "items": {
        "$ref": "#/definitions/Release"
      }
    },
    "system_info": {
      "description": "Info about the toolchain used to build this announcement\n\nDEPRECATED: never appears anymore",
      "anyOf": [
        {
          "$ref": "#/definitions/SystemInfo"
        },
        {
          "type": "null"
        }
      ]
    },
    "systems": {
      "description": "The systems that artifacts were built on",
      "type": "object",
      "additionalProperties": {
        "$ref": "#/definitions/SystemInfo"
      }
    },
    "upload_files": {
      "description": "Files to upload",
      "default": [],
      "type": "array",
      "items": {
        "type": "string"
      }
    }
  },
  "definitions": {
    "Artifact": {
      "description": "A distributable artifact that's part of a Release\n\ni.e. a zip or installer",
      "type": "object",
      "oneOf": [
        {
          "description": "A zip or a tarball",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "executable-zip"
              ]
            }
          }
        },
        {
          "description": "Standalone Symbols/Debuginfo for a build",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "symbols"
              ]
            }
          }
        },
        {
          "description": "Installer",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "installer"
              ]
            }
          }
        },
        {
          "description": "A checksum of another artifact",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "checksum"
              ]
            }
          }
        },
        {
          "description": "The checksums of many artifacts",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "unified-checksum"
              ]
            }
          }
        },
        {
          "description": "A tarball containing the source code",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "source-tarball"
              ]
            }
          }
        },
        {
          "description": "Some form of extra artifact produced by a sidecar build",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "extra-artifact"
              ]
            }
          }
        },
        {
          "description": "An updater executable",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "updater"
              ]
            }
          }
        },
        {
          "description": "A file that already exists",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "sbom"
              ]
            }
          }
        },
        {
          "description": "An OmniBOR Artifact ID",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "omnibor-artifact-id"
              ]
            }
          }
        },
        {
          "description": "Unknown to this version of cargo-dist-schema\n\nThis is a fallback for forward/backward-compat",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "unknown"
              ]
            }
          }
        }
      ],
      "properties": {
        "assets": {
          "description": "Assets included in the bundle (like executables and READMEs)",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Asset"
          }
        },
        "checksum": {
          "description": "id of an Artifact that contains the checksum for this Artifact",
          "type": [
            "string",
            "null"
          ]
        },
        "checksums": {
          "description": "checksums for this artifact\n\nkeys are the name of an algorithm like \"sha256\" or \"sha512\" values are the actual hex string of the checksum",
          "type": "object",
          "additionalProperties": {
            "type": "string"
          }
        },
        "description": {
          "description": "A brief description of what this artifact is",
          "type": [
            "string",
            "null"
          ]
        },
        "install_hint": {
          "description": "A string describing how to install this",
          "type": [
            "string",
            "null"
          ]
        },
        "name": {
          "description": "The unique name of the artifact (e.g. `myapp-v1.0.0-x86_64-pc-windows-msvc.zip`)\n\nIf this is missing then that indicates the artifact is purely informative and has no physical files associated with it. This may be used (in the future) to e.g. indicate you can install the application with `cargo install` or `npm install`.",
          "type": [
            "string",
            "null"
          ]
        },
        "path": {
          "description": "The location of the artifact on the local system",
          "type": [
            "string",
            "null"
          ]
        },
        "target_triples": {
          "description": "The target triple of the bundle",
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    },
    "ArtifactSet": {
      "description": "A handle for talking about ArtifactSets",
      "type": "object",
      "required": [
        "package",
        "public_id"
      ],
      "properties": {
        "announce_url": {
          "type": [
            "string",
            "null"
          ]
        },
        "package": {
          "type": "string"
        },
        "public_id": {
          "type": "string"
        },
        "release_url": {
          "type": [
            "string",
            "null"
          ]
        },
        "set_download_url": {
          "type": [
            "string",
            "null"
          ]
        },
        "upload_url": {
          "type": [
            "string",
            "null"
          ]
        }
      }
    },
    "Asset": {
      "description": "An asset contained in an artifact (executable, license, etc.)",
      "type": "object",
      "oneOf": [
        {
          "description": "An executable artifact",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "executable"
              ]
            },
            "symbols_artifact": {
              "description": "The name of the Artifact containing symbols for this executable",
              "type": [
                "string",
                "null"
              ]
            }
          }
        },
        {
          "description": "A C dynamic library",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "c_dynamic_library"
              ]
            },
            "symbols_artifact": {
              "description": "The name of the Artifact containing symbols for this library",
              "type": [
                "string",
                "null"
              ]
            }
          }
        },
        {
          "description": "A C static library",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "c_static_library"
              ]
            },
            "symbols_artifact": {
              "description": "The name of the Artifact containing symbols for this library",
              "type": [
                "string",
                "null"
              ]
            }
          }
        },
        {
          "description": "A README file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "readme"
              ]
            }
          }
        },
        {
          "description": "A LICENSE file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "license"
              ]
            }
          }
        },
        {
          "description": "A CHANGELOG or RELEASES file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "changelog"
              ]
            }
          }
        },
        {
          "description": "Unknown to this version of cargo-dist-schema\n\nThis is a fallback for forward/backward-compat",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "unknown"
              ]
            }
          }
        }
      ],
      "properties": {
        "id": {
          "description": "A unique opaque id for an Asset",
          "type": [
            "string",
            "null"
          ]
        },
        "name": {
          "description": "The high-level name of the asset",
          "type": [
            "string",
            "null"
          ]
        },
        "path": {
          "description": "The path of the asset relative to the root of the artifact",
          "type": [
            "string",
            "null"
          ]
        }
      }
    },
    "AssetInfo": {
      "description": "Info about an Asset (binary)",
      "type": "object",
      "required": [
        "id",
        "name",
        "system",
        "target_triples"
      ],
      "properties": {
        "id": {
          "description": "unique id of the Asset",
          "type": "string"
        },
        "linkage": {
          "description": "the linkage of this Asset",
          "anyOf": [
            {
              "$ref": "#/definitions/Linkage"
            },
            {
              "type": "null"
            }
          ]
        },
        "name": {
          "description": "filename of the Asset",
          "type": "string"
        },
        "system": {
          "description": "the system it was built on",
          "type": "string"
        },
        "target_triples": {
          "description": "rust-style target triples the Asset natively supports\n\n* length 0: not a meaningful question, maybe some static file * length 1: typical of binaries * length 2+: some kind of universal binary",
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    },
    "BuildEnvironment": {
      "description": "Information about the build environment on this system",
      "oneOf": [
        {
          "description": "Linux-specific information",
          "type": "object",
          "required": [
            "linux"
          ],
          "properties": {
            "linux": {
              "type": "object",
              "properties": {
                "glibc_version": {
                  "description": "The builder's glibc version, relevant to glibc-based builds.",
                  "anyOf": [
                    {
                      "$ref": "#/definitions/GlibcVersion"
                    },
                    {
                      "type": "null"
                    }
                  ]
                }
              }
            }
          },
          "additionalProperties": false
        },
        {
          "description": "macOS-specific information",
          "type": "object",
          "required": [
            "macos"
          ],
          "properties": {
            "macos": {
              "type": "object",
              "required": [
                "os_version"
              ],
              "properties": {
                "os_version": {
                  "description": "The version of macOS used by the builder",
                  "type": "string"
                }
              }
            }
          },
          "additionalProperties": false
        },
        {
          "description": "Windows-specific information",
          "type": "string",
          "enum": [
            "windows"
          ]
        },
        {
          "description": "Unable to determine what the host OS was - error?",
          "type": "string",
          "enum": [
            "indeterminate"
          ]
        }
      ]
    },
    "CiInfo": {
      "description": "CI backend info",
      "type": "object",
      "properties": {
        "github": {
          "description": "GitHub CI backend",
          "anyOf": [
            {
              "$ref": "#/definitions/GithubCiInfo"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "ContainerConfig": {
      "description": "GitHub config that's common between different kinds of jobs (global, local)",
      "type": "object",
      "required": [
        "host",
        "image"
      ],
      "properties": {
        "host": {
          "description": "The host triple of the container, something like `x86_64-unknown-linux-gnu` or `aarch64-unknown-linux-musl` or whatever.",
          "type": "string"
        },
        "image": {
          "description": "The container image to run, something like `ubuntu:22.04` or `quay.io/pypa/manylinux_2_28_x86_64`",
          "type": "string"
        },
        "package_manager": {
          "description": "The package manager to use within the container, like `apt`.",
          "anyOf": [
            {
              "$ref": "#/definitions/PackageManager"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "EnvironmentVariables": {
      "description": "Release-specific environment variables",
      "type": "object",
      "required": [
        "disable_update_env_var",
        "ghe_base_url_env_var",
        "github_base_url_env_var",
        "install_dir_env_var",
        "no_modify_path_env_var",
        "unmanaged_dir_env_var"
      ],
      "properties": {
        "disable_update_env_var": {
          "description": "Environment variable to disable updater features",
          "type": "string"
        },
        "ghe_base_url_env_var": {
          "description": "Environment variable to set the GitHub Enterprise base URL",
          "type": "string"
        },
        "github_base_url_env_var": {
          "description": "Environment variable to set the GitHub base URL",
          "type": "string"
        },
        "install_dir_env_var": {
          "description": "Environment variable to force an install location",
          "type": "string"
        },
        "no_modify_path_env_var": {
          "description": "Environment variable to disable modifying the path",
          "type": "string"
        },
        "unmanaged_dir_env_var": {
          "description": "Environment variable to force an unmanaged install location",
          "type": "string"
        }
      }
    },
    "GhaRunStep": {
      "description": "A GitHub Actions \"run\" step, either bash or powershell",
      "oneOf": [
        {
          "description": "see [`DashScript`]",
          "type": "object",
          "required": [
            "run",
            "shell"
          ],
          "properties": {
            "run": {
              "type": "string"
            },
            "shell": {
              "type": "string",
              "enum": [
                "sh"
              ]
            }
          }
        },
        {
          "description": "see [`PowershellScript`]",
          "type": "object",
          "required": [
            "run",
            "shell"
          ],
          "properties": {
            "run": {
              "type": "string"
            },
            "shell": {
              "type": "string",
              "enum": [
                "pwsh"
              ]
            }
          }
        }
      ]
    },
    "GithubCiInfo": {
      "description": "Github CI backend",
      "type": "object",
      "properties": {
        "artifacts_matrix": {
          "description": "Github CI Matrix for upload-artifacts",
          "anyOf": [
            {
              "$ref": "#/definitions/GithubMatrix"
            },
            {
              "type": "null"
            }
          ]
        },
        "external_repo_commit": {
          "description": "A specific commit to tag in an external repository",
          "type": [
            "string",
            "null"
          ]
        },
        "pr_run_mode": {
          "description": "What kind of job to run on pull request",
          "anyOf": [
            {
              "$ref": "#/definitions/PrRunMode"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "GithubHosting": {
      "description": "Github Hosting",
      "type": "object",
      "required": [
        "artifact_base_url",
        "artifact_download_path",
        "owner",
        "repo"
      ],
      "properties": {
        "artifact_base_url": {
          "description": "The URL of the host for GitHub, usually `\"https://github.com\"` (This can vary for GitHub Enterprise)",
          "type": "string"
        },
        "artifact_download_path": {
          "description": "The path of the release without the base URL\n\ne.g. `/myowner/myrepo/releases/download/v1.0.0/`",
          "type": "string"
        },
        "owner": {
          "description": "The owner of the repo",
          "type": "string"
        },
        "repo": {
          "description": "The name of the repo",
          "type": "string"
        }
      }
    },
    "GithubLocalJobConfig": {
      "description": "Used in `github/release.yml.j2` to template out \"local\" build jobs",
      "type": "object",
      "required": [
        "dist_args",
        "host",
        "install_dist",
        "runner"
      ],
      "properties": {
        "cache_provider": {
          "description": "What cache provider to use",
          "type": [
            "string",
            "null"
          ]
        },
        "container": {
          "description": "Container image to run the job in, using GitHub's builtin container support, see <https://docs.github.com/en/actions/writing-workflows/choosing-where-your-workflow-runs/running-jobs-in-a-container>\n\nThis doesn't allow mounting volumes, or anything, because we're only able to set the `container` key to something stringy",
          "anyOf": [
            {
              "$ref": "#/definitions/ContainerConfig"
            },
            {
              "type": "null"
            }
          ]
        },
        "dist_args": {
          "description": "Arguments to pass to dist",
          "type": "string"
        },
        "host": {
          "description": "Host triple of the runner (well-known, custom, or best guess). If the runner is one of GitHub's official runner images, the platform is hardcoded. If it's custom, then we have a `target_triple => runner` in the config",
          "type": "string"
        },
        "install_cargo_auditable": {
          "description": "Expression to execute to install cargo-auditable",
          "anyOf": [
            {
              "$ref": "#/definitions/GhaRunStep"
            },
            {
              "type": "null"
            }
          ]
        },
        "install_dist": {
          "description": "Expression to execute to install dist",
          "allOf": [
            {
              "$ref": "#/definitions/GhaRunStep"
            }
          ]
        },
        "install_omnibor": {
          "description": "Expression to execute to install omnibor-cli",
          "anyOf": [
            {
              "$ref": "#/definitions/GhaRunStep"
            },
            {
              "type": "null"
            }
          ]
        },
        "packages_install": {
          "description": "Command to run to install dependencies",
          "type": [
            "string",
            "null"
          ]
        },
        "runner": {
          "description": "GHA's `runs-on` key: Github Runner image to use, see <https://github.com/actions/runner-images> and <https://docs.github.com/en/actions/writing-workflows/choosing-where-your-workflow-runs/choosing-the-runner-for-a-job>\n\nThis is not necessarily a well-known runner, it could be something self-hosted, it could be from BuildJet, Namespace, etc.",
          "type": "string"
        },
        "targets": {
          "description": "Target triples to build for",
          "type": [
            "array",
            "null"
          ],
          "items": {
            "type": "string"
          }
        }
      }
    },
    "GithubMatrix": {
      "description": "Github CI Matrix",
      "type": "object",
      "properties": {
        "include": {
          "description": "define each task manually rather than doing cross-product stuff",
          "type": "array",
          "items": {
            "$ref": "#/definitions/GithubLocalJobConfig"
          }
        }
      }
    },
    "GlibcVersion": {
      "description": "Minimum glibc version required to run software",
      "type": "object",
      "required": [
        "major",
        "series"
      ],
      "properties": {
        "major": {
          "description": "Major version",
          "type": "integer",
          "format": "uint64",
          "minimum": 0.0
        },
        "series": {
          "description": "Series (minor) version",
          "type": "integer",
          "format": "uint64",
          "minimum": 0.0
        }
      }
    },
    "Hosting": {
      "description": "Possible hosting providers",
      "type": "object",
      "properties": {
        "axodotdev": {
          "description": "Hosted on Axo Releases",
          "anyOf": [
            {
              "$ref": "#/definitions/ArtifactSet"
            },
            {
              "type": "null"
            }
          ]
        },
        "github": {
          "description": "Hosted on Github Releases",
          "anyOf": [
            {
              "$ref": "#/definitions/GithubHosting"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "Library": {
      "description": "Represents a dynamic library located somewhere on the system",
      "type": "object",
      "required": [
        "path"
      ],
      "properties": {
        "package_manager": {
          "description": "Which package manager provided this library",
          "anyOf": [
            {
              "$ref": "#/definitions/PackageManager"
            },
            {
              "type": "null"
            }
          ]
        },
        "path": {
          "description": "The path to the library; on platforms without that information, it will be a basename instead",
          "type": "string"
        },
        "source": {
          "description": "The package from which a library comes, if relevant",
          "type": [
            "string",
            "null"
          ]
        }
      }
    },
    "Linkage": {
      "description": "Information about dynamic libraries used by a binary",
      "type": "object",
      "properties": {
        "frameworks": {
          "description": "Frameworks, only used on macOS",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Library"
          },
          "uniqueItems": true
        },
        "homebrew": {
          "description": "Libraries provided by the Homebrew package manager",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Library"
          },
          "uniqueItems": true
        },
        "other": {
          "description": "Libraries which don't fall into any other categories",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Library"
          },
          "uniqueItems": true
        },
        "public_unmanaged": {
          "description": "Public libraries not provided by the system and not managed by any package manager",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Library"
          },
          "uniqueItems": true
        },
        "system": {
          "description": "Libraries included with the operating system",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Library"
          },
          "uniqueItems": true
        }
      }
    },
    "PackageManager": {
      "description": "Represents the package manager a library was installed by",
      "oneOf": [
        {
          "description": "Homebrew (usually for Mac)",
          "type": "string",
          "enum": [
            "homebrew"
          ]
        },
        {
          "description": "Apt (Debian, Ubuntu, etc)",
          "type": "string",
          "enum": [
            "apt"
          ]
        }
      ]
    },
    "PrRunMode": {
      "description": "Type of job to run on pull request",
      "oneOf": [
        {
          "description": "Do not run on pull requests at all",
          "type": "string",
          "enum": [
            "skip"
          ]
        },
        {
          "description": "Only run the plan step",
          "type": "string",
          "enum": [
            "plan"
          ]
        },
        {
          "description": "Build and upload artifacts",
          "type": "string",
          "enum": [
            "upload"
          ]
        }
      ]
    },
    "Release": {
      "description": "A Release of an Application",
      "type": "object",
      "required": [
        "app_name",
        "app_version"
      ],
      "properties": {
        "app_name": {
          "description": "The name of the app",
          "type": "string"
        },
        "app_version": {
          "description": "The version of the app",
          "type": "string"
        },
        "artifacts": {
          "description": "The artifacts for this release (zips, debuginfo, metadata...)",
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "display": {
          "description": "Whether to advertise this app's installers/artifacts in announcements",
          "type": [
            "boolean",
            "null"
          ]
        },
        "display_name": {
          "description": "Alternative display name that can be prettier",
          "type": [
            "string",
            "null"
          ]
        },
        "env": {
          "description": "Environment variables which control this release's installer's behaviour",
          "anyOf": [
            {
              "$ref": "#/definitions/EnvironmentVariables"
            },
            {
              "type": "null"
            }
          ]
        },
        "hosting": {
          "description": "Hosting info",
          "allOf": [
            {
              "$ref": "#/definitions/Hosting"
            }
          ]
        }
      }
    },
    "SystemInfo": {
      "description": "Info about a system used to build this announcement.",
      "type": "object",
      "required": [
        "build_environment",
        "id"
      ],
      "properties": {
        "build_environment": {
          "description": "Environment of the System",
          "allOf": [
            {
              "$ref": "#/definitions/BuildEnvironment"
            }
          ]
        },
        "cargo_version_line": {
          "description": "The version of Cargo used (first line of cargo -vV)",
          "type": [
            "string",
            "null"
          ]
        },
        "id": {
          "description": "The unique id of the System",
          "type": "string"
        }
      }
    }
  }
}
